docs/proposals/deployment.md

```diff
@@ -0,0 +1,254 @@
+4. Scale up the new RC and scale down the olds ones as per the DeploymentStrategy.
+   - Raise an event if we detect an error, like new pods failing to come up.
+5. Go back to step 1 unless the new RC has been ramped up to desired replicas
+   and the old RCs have been ramped down to 0.
+6. Cleanup.
+
+DeploymentController is stateless so that it can recover incase it crashes during a deployment.
+
+## Changing Deployment mid-way
+
+### Updating
+
+Users can update an ongoing deployment before it is completed.
+In this case, the existing deployment will be stalled and the new one will
+begin.
+For ex: consider the following case:
+- User creates a deployment to rolling-update 10 pods with image:v1 to
+  pods with image:v2.
+- User then updates this deployment to create pods with image:v3,
+  when the image:v2 RC had been ramped up to 5 pods and the image:v1 RC
+  had been ramped down to 5 pods.
+- When Deployment Controller observes the new deployment, it will create
+  a new RC for creating pods with image:v3. It will then start ramping up this
+  new RC to 10 pods and will ramp down both the existing RCs to 0.
+
+### Deleting
+
+Users can pause/cancel a deployment by deleting it before it is completed.
+Recreating the same deployment will resume it.
+For ex: consider the following case:
+- User creates a deployment to rolling-update 10 pods with image:v1 to
+  pods with image:v2.
+- User then deletes this deployment while the old and new RCs are at 5 replicas each.
+  User will end up with 2 RCs with 5 replicas each.
+User can then create the same deployment again in which case, DeploymentController will
+notice that the second RC exists already which it can ramp up while ramping down
+the first one.
+
+### Rollback
+
+We want to allow the user to rollback a deployment. To rollback a
+completed (or ongoing) deployment, user can create (or update) a deployment with
+DeploymentSpec.PodTemplateSpec = oldRC.PodTemplateSpec.
+
+## Deployment Strategies
+
+DeploymentStrategy specifies how the new RC should replace existing RCs.
+To begin with, we will support 2 types of deployment:
+* Recreate: We kill all existing RCs and then bring up the new one. This results
+  in quick deployment but there is a downtime when old pods are down but
+  the new ones have not come up yet.
+* Rolling update: We gradually scale down old RCs while scaling up the new one.
+  This results in a slower deployment, but there is no downtime. At all times
+  during the deployment, there are a few pods available (old or new). The number
+  of available pods and when is a pod considered "available" can be configured
+  using RollingUpdateDeploymentSpec.
+
+In future, we want to support more deployment types.
+
+## Future
+
+Apart from the above, we want to add support for the following:
+* Running the deployment process in a pod: In future, we can run the deployment process in a pod. Then users can define their own custom deployments and we can run it using the image name.
```

Comment:
As @bgrant0607 clarified, Deployment is more like a ReplicationController rather than a Job. It is a continuously running process, trying to bring the system state towards the desired state. It wont fail (can exponentially back off) or succeed.

In light of this info, DeploymentController wont be able to use rolling_updater as is. 
We will need some refactoring to share common code. For ex: Deployment can use the when to scale up and scale down logic in the for loop in rolling_updater.